Open Judge System/OJS.Services/Business/ProblemGroups/ProblemGroupsBusinessService.cs

```diff
@@ -10,26 +12,52 @@ public class ProblemGroupsBusinessService : IProblemGroupsBusinessService
         //TODO: Add to Resource
         private const string CannotDeleteProblemGroupWithProblems = "Cannot delete problem group with problems";
 
+        private readonly IEfDeletableEntityRepository<ProblemGroup> problemGroups;
         private readonly IProblemGroupsDataService problemGroupsData;
 
-        public ProblemGroupsBusinessService(IProblemGroupsDataService problemGroupsData) =>
+        public ProblemGroupsBusinessService(
+            IEfDeletableEntityRepository<ProblemGroup> problemGroups,
+            IProblemGroupsDataService problemGroupsData)
+        {
+            this.problemGroups = problemGroups;
             this.problemGroupsData = problemGroupsData;
+        }
 
         public ServiceResult DeleteById(int id)
         {
             var problemGroup = this.problemGroupsData.GetById(id);
 
             if (problemGroup != null)
             {
-                if (!problemGroup.Problems.All(p => p.IsDeleted))
+                if (problemGroup.Problems.Any(p => !p.IsDeleted))
                 {
                     return new ServiceResult(CannotDeleteProblemGroupWithProblems);
                 }
 
-                this.problemGroupsData.Delete(problemGroup);
+                this.Delete(problemGroup);
             }
 
             return ServiceResult.Success;
         }
+
+        public void DeleteByContest(int contestId)
+        {
+            var problemGroupIds = this.problemGroupsData
+                .GetAllByContest(contestId)
+                .Where(pg => !pg.IsDeleted && pg.Problems.All(p => p.IsDeleted))
+                .Select(pg => pg.Id)
+                .ToList();
+
+            if (problemGroupIds.Any())
```

Comment:
What's the point of making 2 queries, just Delete by the entire filter expression, instead of filtering, materializing and then deleting by filtering the ids.